module/VuFind/src/VuFind/Service/Feature/RetryTrait.php

```diff
@@ -0,0 +1,127 @@
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development
+ */
+namespace VuFind\Service\Feature;
+
+/**
+ * Trait that provides support for calling a method with configurable retries
+ *
+ * @category VuFind
+ * @package  Service
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development
+ */
+trait RetryTrait
+{
+    /**
+     * Retry options
+     *
+     * @var array
+     */
+    protected $retryOptions = [
+        'retryCount' => 5,            // number of retries (set to 0 to disable)
+        'firstBackoff' => 0,          // backoff (delay) before first retry
+                                      // (milliseconds)
+        'subsequentBackoff' => 200,   // backoff (delay) before subsequent retries
+                                      // (milliseconds)
+        'exponentialBackoff' => true, // whether to use exponential backoff
+        'maximumBackoff' => 1000      // maximum backoff (milliseconds)
+    ];
+
+    /**
+     * Call a method and retry the call if an exception is thrown
+     *
+     * @param callable  $callback       Method to call
+     * @param ?callable $statusCallback Status callback called before retry and after
+     * a successful retry
+     * @param array     $options        Optional options to override defaults in
+     * $this->retryOptions
+     *
+     * @return mixed
+     */
+    protected function callWithRetry(
+        callable $callback,
+        ?callable $statusCallback = null,
+        array $options = []
+    ) {
+        $attempt = 0;
+        $exception = null;
+        $options = array_merge($this->retryOptions, $options);
+        do {
+            if (++$attempt > 1 && $statusCallback) {
+                $statusCallback($attempt, false);
+            }
+            try {
+                $result = $callback();
+                if ($attempt > 1 && $statusCallback) {
+                    $statusCallback($attempt, true);
+                }
+                return $result;
+            } catch (\Exception $e) {
+                if (null === $exception) {
+                    $exception = $e;
+                }
+                if ($checkCallback = $options['retryableExceptionCallback'] ?? '') {
+                    if (!$checkCallback($e)) {
+                        break;
+                    }
+                }
+            }
+        } while ($this->checkRetryAndSleep($attempt, $options));
+        // No success, re-throw the original exception:
+        throw $exception;
+    }
+
+    /**
+     * Check if the call needs to be retried and sleep before the retry
+     *
+     * @param int   $attempt Failed attempt number
+     * @param array $options Current options
+     *
+     * @return bool
+     */
+    protected function checkRetryAndSleep(int $attempt, array $options): bool
```

Comment:
Maybe it would make sense to refactor this so that there's a separate getSleepDurationForAttempt method, and then we just pass that value to usleep. That would probably allow you to make your test a little more straightforward (you could test that sleep durations are calculated correctly without having to actually sleep and measure the duration), and it might slightly improve readability here.